pkg/controller/sls/store_controller.go

```diff
@@ -0,0 +1,192 @@
+/*
+
+ Copyright 2021 The Crossplane Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+
+*/
+
+package sls
+
+import (
+	"context"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	slsv1alpha1 "github.com/crossplane/provider-alibaba/apis/sls/v1alpha1"
+	"github.com/crossplane/provider-alibaba/apis/v1alpha1"
+	slsclient "github.com/crossplane/provider-alibaba/pkg/clients/sls"
+)
+
+const (
+	errNotStore = "managed resource is not a SLS store custom resource"
+)
+
+// SetupStore adds a controller that reconciles SLSStores.
+func SetupStore(mgr ctrl.Manager, l logging.Logger) error {
+	name := managed.ControllerName(slsv1alpha1.StoreGroupKind)
+	options := []managed.ReconcilerOption{
+		managed.WithExternalConnecter(&logStoreConnector{
+			client:      mgr.GetClient(),
+			usage:       resource.NewProviderConfigUsageTracker(mgr.GetClient(), &v1alpha1.ProviderConfigUsage{}),
+			NewClientFn: slsclient.NewClient,
+		}),
+		managed.WithLogger(l.WithValues("controller", name)),
+		managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))),
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&slsv1alpha1.Store{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(slsv1alpha1.StoreGroupVersionKind), options...))
+}
+
+type logStoreConnector struct {
+	client      client.Client
+	usage       resource.Tracker
+	NewClientFn func(accessKeyID, accessKeySecret, region string) *slsclient.LogClient
+}
+
+func (c *logStoreConnector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) { //nolint:gocyclo
+	cr, ok := mg.(*slsv1alpha1.Store)
+	if !ok {
+		return nil, errors.New(errNotStore)
+	}
+
+	var (
+		sel    *xpv1.SecretKeySelector
+		region string
+	)
+
+	switch {
```

Comment:
It looks like this is being replicated in a number of controllers, might make sense to move it to a common package, such as `client`. `crossplane-runtime` also supports a common credentials extraction mechanism now that will allow you to fetch credentials from a variety of sources. Example here: https://github.com/crossplane/provider-aws/blob/3243f6c4cfa8c860ceee705c9eca72ea68398545/pkg/clients/aws.go#L105

Looks like there might need to be a crossplane-runtime update here first though so might make sense to tackle in a follow-up.

Reply:
I see this has already been implemented in https://github.com/crossplane/provider-alibaba/pull/64, apologies for the delayed review there! I believe it needs a rebase but I will prioritize getting eyes on that!